src/lib.rs

```diff
@@ -0,0 +1,1166 @@
+//
+// This license is subject to the following condition: The above copyright notice and either this
+// complete permission notice or at a minimum a reference to the UPL must be included in all copies
+// or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+//! A vector of bits ("Vob") is a sequence of bits which exposes a `Vec`-like interface. Whereas
+//! `Vec<bool>` requires 1 byte of storage per bit, `Vob` requires only 1 bit of storage per bit.
+//!
+//! The main documentation for this crate can be found in the [`Vob`](struct.Vob.html) struct.
+
+extern crate num_traits;
+
+use std::cmp::PartialEq;
+use std::fmt;
+use std::fmt::Debug;
+use std::hash::{Hash, Hasher};
+use std::iter::FromIterator;
+use std::mem::size_of;
+use std::ops::{Index, Range};
+use std::slice;
+
+use num_traits::{PrimInt, One, Zero};
+
+
+/// A Vob is a "vector of bits": a sequence of bits which exposes a `Vec`-like interface. Whereas
+/// `Vec<bool>` requires 1 byte of storage per bit, `Vob` requires only 1 bit of storage per bit.
+/// For larger numbers of bits, Vobs can lead to a significant memory decrease and performance
+/// increase.
+///
+/// # Examples
+/// The `vob!` macro makes creating small `Vob`s easy:
+///
+/// ```rust
+/// # #[macro_use] extern crate vob;
+/// # fn main() {
+/// let mut v = vob![true, false, true];
+/// assert_eq!(v[1], false);
+/// # }
+/// ```
+///
+/// Operations such as `and`ing two `Vob`s together are quick; one can also quickly identify which
+/// bits are set:
+///
+/// ```rust
+/// use vob::Vob;
+/// let mut v1 = Vob::from_elem(256, false);
+/// let mut v2 = Vob::from_elem(256, false);
+/// v1.set(67, true);
+/// v2.set(67, true);
+/// v1.set(188, true);
+/// v1.and(&v2);
+/// let num_bits_set = v1.iter_set_bits().count();
+/// assert_eq!(num_bits_set, 1);
+/// ```
+///
+///
+/// ## Storage backing type
+///
+/// `Vob`s default to using `usize` as a storage backing type. This is generally a substantial win
+/// over using smaller storage types if you use functions such as
+/// [`or()`](struct.Vob.html#method.or). In such cases, `usize` on a 64-bit machine is almost
+/// exactly twice as fast as using `u32`. If you only ever set and get individual bits, a smaller
+/// data type might be marginally more effective: for such use cases `u32` is around 1% faster than
+/// `usize` on a 64-bit machine. You can choose your own storage type with the
+/// [`new_with_storage_type()`](struct.Vob.html#method.new_with_storage_type) constructor. In
```

Comment:
Unfortunately Rust's default types don't really work as one would expect :( This means we end up with two `impl` blocks. The first `impl` block handles the "make `usize` the default" which contains a `new` function. The second `impl` block can't overload `new`, so has to use this awkward name.

In an ideal world we'd have had one `impl` block specified thus:

```rust
impl<T: Debug + PrimInt + One + Zero> Vob<T=usize> {
```

Unfortunately that just doesn't work in `rustc` at the moment. AFAICS the default types are a half-implemented hack...

Reply:
I wonder if it would be better to force the user to specify a backing type rather than this `new_with_storage_type()` dance.

At first I was all like "huh, I wonder what storage type `0` is?", then i realised that it's just a workaround to allow us specify a different backing type and `0` is the capacity.

If we force the user to specify a type, then if and when Rust eventually allows inference on `new()`, it's a no-op for consumers to use the new interface. i.e. `let v: Vob<u8> = Vob::new()` will work now and when Rust is fixed.

Reply:
No, we *definitely* don't want them to specify it -- people will often choose an inappropriate type (for a good approximation of the truth `usize` is, AFAICS, always the right choice; but in case I'm wrong, I provide a backdoor). At some point Rust will fix default generic types and we can deprecate this method.

Reply:
Right OK. It's not a nice interface :(

Reply:
What I could do (indeed, an earlier version did have) is a disclaimer on `new_with_storage_type` which says "unless you have a good reason for doing so, use `new` or `with_capacity`). [I worried this was a bit too pedantic, but maybe I was wrong.]

Reply:
[I also sort-of disagree that it's not a good interface: the main method `new` (and `with_capacity`) "do the right thing (TM)", but there's a backdoor (`new_with_storage_type`) forced upon us by a lack of functionality in rustc. The latter bit isn't ideal, but I wouldn't say it's bad. Certainly I've written many worse APIs, even in the last week!]

Reply:
To be clear. I think you've done the best with the limitations of the tools.

I'd also have accepted forcing a decision from the user, but we've already discussed that.